smarts/core/utils/traffic_history_serivce.py

```diff
@@ -0,0 +1,153 @@
+    memory use of traffic history data
+    """
+
+    def __init__(self, history_file_path):
+        self._history_file_path = history_file_path
+        self._all_timesteps = set()
+        self._current_traffic_history = {}
+        self._prev_batch_history = {}
+        # return if traffic history is not used
+        if history_file_path is None:
+            return
+
+        send_data_conn, receive_data_conn = Pipe()
+        self._receive_data_conn = receive_data_conn
+        self._request_queue = Queue()
+        self._fetch_history_proc = Process(
+            target=self._fetch_history,
+            args=(
+                send_data_conn,
+                self._request_queue,
+                self._history_file_path,
+                len(self._all_timesteps),
+            ),
+        )
+        self._fetch_history_proc.daemon = True
+        self._fetch_history_proc.start()
+
+        self._range_start = 0
+        self._batch_size = 300
+        # initialize
+        with open(self._history_file_path, "rb") as f:
+            for index, (t, vehicles_state) in enumerate(
+                ijson.kvitems(f, "", use_float=True)
+            ):
+                self._all_timesteps.add(t)
+                if (
+                    self._range_start <= index
+                    and index <= self._range_start + self._batch_size
+                ):
+                    self._current_traffic_history[t] = vehicles_state
+        self._range_start += self._batch_size
+        # prepares the next batch
+        self._prepare_next_batch()
+        self._receive_data_conn.recv()
+
+    def _fetch_history(
+        self, send_data, request_queue, history_file_path, total_timestep_size
+    ):
+        """prepare 1 batch ahead, when received request, immediately return the previously
+        prepared batch and prepares the next batch.
+        """
+        return_batch = {}
+        while True:
+            historyRange = request_queue.get()
+            assert isinstance(historyRange, RequestHistoryRange)
+            send_data.send(return_batch)
+            return_batch = {}
+            with open(history_file_path, "rb") as f:
+                for index, (t, vehicles_state) in enumerate(
+                    ijson.kvitems(f, "", use_float=True)
+                ):
+                    if (
+                        historyRange.start_index <= index
+                        and index <= historyRange.end_index
+                    ):
+                        return_batch[t] = vehicles_state
+        send_data.close()
+
+    @property
+    def all_timesteps(self):
+        return self._all_timesteps
+
+    @property
+    def traffic_history(self):
+        return {**self._current_traffic_history, **self._prev_batch_history}
+
+    def _prepare_next_batch(self):
+        self._request_queue.put(
+            RequestHistoryRange(
+                start_index=self._range_start,
+                end_index=self._range_start + self._batch_size,
```

Comment:
I might replace this end index with count. Otherwise it will look like the following to get the batch size intended:

```suggestion
                end_index=self._range_start + self._batch_size - 1,
```